File: stella_fuzzy_pattern_matcher/SFPM-JS/src/Rule.js
// src/Rule.js

export class Rule {
    /**
     * @param {import('./Criteria').Criteria[]} criterias
     * @param {(...data) => void} payload
     * @param {string} [name='']
     * @param {number} [priority=0]
     */
    constructor(criterias, payload, name = "", priority = 0) {
        this.criterias = criterias;
        this.payload = payload;
        this.name = name;
        this.priority = priority;
    }

    get criteriaCount() {
        return this.criterias.length;
    }

    /**
     * @param {import('./FactSource').FactSource} facts
     * @returns {{isTrue: boolean, matchedCriteriaCount: number}}
     */
    evaluate(facts) {
        if (this.criteriaCount === 0) {
            return { isTrue: true, matchedCriteriaCount: 0 };
        }

        for (const criteria of this.criterias) {
            if (!criteria.evaluate(facts)) {
                return { isTrue: false, matchedCriteriaCount: 0 };
            }
        }

        return { isTrue: true, matchedCriteriaCount: this.criteriaCount };
    }

    executePayload(...data) {
        this.payload(...data);
    }
}

File: stella_fuzzy_pattern_matcher/SFPM-PY/SFPM/rule.py
from icontract import require, ensure

class Rule():
    def __init__(self, priority = 0, criterias = None, payload = None):
        self.priority = priority
        self.criterias = criterias
        self.payload = payload

    def execute(self, *args):
        self.payload(*args);
    
    def count(self):
        """Returns the number of criterias the rule has"""
        return len(self.criterias)

    @require(lambda self, facts: self.criterias, "Criterias cannot be an empty list")
    @ensure(lambda result: isinstance(result[0], bool))
    def evaluate(self, facts):

        matched_criteria_number = 0
        matched_all_criteria = False
        for criteria in self.criterias:
            result = criteria.evaluate(facts)
            if result == True:
                matched_criteria_number += 1

        if matched_criteria_number == len(self.criterias):
            matched_all_criteria = True

        return matched_all_criteria, matched_criteria_number
